### PR TITLE
Integrate prettier to make the format consistent

### PR DESCRIPTION
**Pull request by nvh95**
# Motivation

- This project comprised many markdown files. However, it lacks a way to make the format consistent across the project.
- Future Pull Requests might mess up lines that they don't edit. That affects the history line.
- Standard formatted markdown can help third-party systems process content from this project seamlessly (e.g: [memos](https://github.com/thien-do/memos-pub)).

# How

- [x] Integrate [prettier](https://prettier.io/) to `backend-swe-interview-questions` - a de facto tool for formating code.
- [x] Add a GitHub action to indicate if a pushed commit/ a Pull Request have inconsistency format.

# Pending Items
- [x] Since we need to reformat the existing markdown files. @tamhoang1412 should be the one to do this to preserve the git history.

### CODE REVIEW

File: package.json
Comment (huytd) [inline package.json:18]: Do we need this sort script?

**Comment by huytd** [comment]
I think the PR is pretty good so far. There is one thing I want to bring up for discussion.

Should we enforce the formatting somehow? So that every PR created will be in formatted state. I'm thinking of two approaches:

1. Create a pre-commit hook to run Prettier before the user make any commit. The caveat is, if someone make a PR by editing directly on Github, no formatting will be done.
2. I'm unable to check if the current Github Action will make any commit after Prettier run or not (it only check, right?). But I think it would be great if we make the action run and commit/push the formatted markdown after every PR merge. This will also solve the problem in the first approach.

What do you think? @nvh95 @tamhoang1412

File: package.json
Comment (nvh95) [reply package.json:18]: It's just a trick I recently have. In the future, we may install dependencies by editing `package.json` directly (e.g: integrate `lint-stage`), or update some fields in that file. To make `package.json` organized, just need to run `npm run sort`, it will sort all fields recursively.
Let's think of it as `prettier`, but for `package.json`

**Comment by nvh95** [comment]
@huytd 
> Create a pre-commit hook to run Prettier before the user make any commit. The caveat is, if someone make a PR by editing directly on Github, no formatting will be done.

I think it's a great idea. However, please remember that users clone repository locally have an option to push the changes without installing prettier 😂. So this initiative mitigates the problem, not solves it completely.

> I'm unable to check if the current Github Action will make any commit after Prettier run or not (it only check, right?). But I think it would be great if we make the action run and commit/push the formatted markdown after every PR merge. This will also solve the problem in the first approach.

The current github action does not make any commits. It just check the format (Notice the flag `--check`, not `--write`). I think we can support that. But if we go with that approach, the history line of this project will be filled with reformat commits from GitHub actions, that what I personally don't like.
I think we can just check the format. If it fails, maintainer does not merge the PR but file a request change. We can also add `CONTRIBUTING.md` to guide them how to use prettier to format the project. We can also make a bot 🤖 to reply to PR that they fails CI checking the format and show the link to the guideline (just idea, not know how to do it now, can leave it as room for improvement in next PRs)

File: package.json
Comment (nvh95) [reply package.json:18]: Moreover, it use `npx`, so it does not cost any disk storage to the users (except some bytes for the script 😂😂)

**Comment by nvh95** [comment]
@huytd 
For the pre-commit approach. Besides the pros you mentioned. Following are the cons:
- I personally have bad experience integrating it. It's not a "just work" solution.
- It takes contributors long time to finish commiting (4-5 seconds??). To me, it's a bad DX.
- If users install the [VSCode Prettier extension](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode), pre-commit hook would not be necessary anymore (because that extension format the file on save), just wasting us a few seconds when committing

**Comment by nvh95** [comment]
What do you think @tamhoang1412? I think the discussion is longer than the code itself. 😅

**Comment by huytd** [comment]
😂 

Yeah I think just run the check and let the contributor handle the code format from now on is good enough. Let's merge this.

Btw, we can config the Github Action to commit on behalf of the repository owner, but maybe it's a bit over-engineering 😂

**Review by huytd — APPROVED** [review]

**Comment by nvh95** [comment]
@huytd Yeah. Let's run it for a couple of new PRs to see how it goes and we can decide the next steps. 
Thanks for your comments and feedback.

**Comment by tamhoang1412** [comment]
@all-contributors please add @nvh95 for code

**Pull request merged**
